### ISSUE TITLE
Implement reference support in languageserver [phab:D1759]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2192]
_Created by **djrenren** at 2017-07-05 20:43:01_: 

Implements "Find References" via the Language Server Protocol
by deferring to the Kythe xref service

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-05 21:25:32_:

  - Ran clang-format and cleaned up the result

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-05 22:38:08_:

Please add trailing newlines to all files

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-06 17:47:34_:

This command also needs `mkdir -p src/proto` to work the first time.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-06 17:47:34_:

These tests fail for me:

```
  1) PathContext normalizeLSPath should strip git path extras:
     TypeError: url_1.URL is not a constructor
      at Object.normalizeLSPath (dist/src/pathContext.js:101:17)
      at Context.it (dist/test/pathContext_test.js:54:40)

  2) PathContext normalizeLSPath should strip file protocol:
     TypeError: url_1.URL is not a constructor
      at Object.normalizeLSPath (dist/src/pathContext.js:101:17)
      at Context.it (dist/test/pathContext_test.js:58:40)
```

**Comment by kythe** [comment]
_**schroederc** required changes at 2017-07-06 17:47:34_

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-06 17:49:18_:

Add a section for running the server.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-06 17:56:12_:

Scratch that.  They pass with a newer version of NodeJS.

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-06 20:52:49_:

  - Handle span errors and support all transport types

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-06 21:09:36_:

  - Mention the actual protocol in the docs

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-06 22:38:39_:

  - Fix path mapping errors

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-07 01:06:19_:

  - Allow local patterns for paths

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-07 18:51:50_:

  - Add newlines and defaults for proto values

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-07 19:25:24_:

  - Unescape generated local paths

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-07 19:29:10_:

  - Reset DEFAULT_CONFIG #Kythe4Lyfe

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-07 19:38:42_:

Please add the other non-caller kinds:

```
definition_kind: kythe.proto.CrossReferencesRequest.DefinitionKind.BINDING_DEFINITIONS,
declaration_kind: kythe.proto.CrossReferencesRequest.DeclarationKind.ALL_DECLARATIONS,
```

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-07 20:00:28_:

  - Include bindings and declarations in references

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-10 23:18:13_:

  - Fix src/proto handling in proto npm script

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-07-11 05:08:56_:

nit: protos

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-07-11 05:08:56_:

nit: when the protos change.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-07-11 05:08:56_:

Are ranges inclusive? Should this be `pos.character < range.end.character` instead?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-07-11 05:08:56_:

the...?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-07-11 05:08:56_:

Please add more module-level documentation about the format of these strings, and (more generally) about the problem this is solving. Some worked examples would be very useful.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-07-11 05:08:56_:

Why? Is this related to the path-to-regexp restriction described below?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-07-11 05:08:56_:

Here and elsewhere appropriate: please describe what your various members are used for.

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-11 20:05:36_:

  - Address feedback

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-07-11 21:09:07_:

Consider checking whether the `.git` extension is actually there to guard against failing a bounds check.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-07-11 21:09:07_:

Is there no syntactic sugar for this? It's unfortunate to be writing Go. Are exceptions Bad?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-07-11 21:36:38_:

Deferring this is OK

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2017-07-11 21:36:38_

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-11 22:36:39_:

  - Check for the .git extension of git urls

**Comment by kythe** [comment]
_**djrenren** closed via commit ddb5ada3d5a3cd61868b62211d0e4a3a42e7255f at 2017-07-11 22:37:30_
